### PR TITLE
Template matching with parameters

### PR DESCRIPTION
**Pull request by eggerdj**
### Summary
This PR is to allow template matching with templates that have parameters.

### Details and comments
The template matching is changed so that gates are compared with a `soft_compare` function that ignores parameters.
As example the soft compare does the following on phase gates 
```
PhaseGate(0.5).soft_compare(PhaseGate(Parameter('β'))) -> True
PhaseGate(0.5).soft_compare(PhaseGate(0.25)) -> False
```
The PR also implements a method to make sure that the match is consistent in terms of parameter values. For example, the template
```
             ┌───────────┐                  ┌────────┐
        q_0: ┤ P(-1.0*β) ├──■────────────■──┤0       ├
             ├───────────┤┌─┴─┐┌──────┐┌─┴─┐│  CZ(β) │
        q_1: ┤ P(-1.0*β) ├┤ X ├┤ P(β) ├┤ X ├┤1       ├
             └───────────┘└───┘└──────┘└───┘└────────┘
```
must produce two matches on the circuit shown below and the circuit must remain correct.
```
             ┌───────┐
        q_0: ┤ P(-2) ├──■────────────■─────────────────────────────
             ├───────┤┌─┴─┐┌──────┐┌─┴─┐┌───────┐
        q_1: ┤ P(-2) ├┤ X ├┤ P(2) ├┤ X ├┤ P(-3) ├──■────────────■──
             ├───────┤└───┘└──────┘└───┘└───────┘┌─┴─┐┌──────┐┌─┴─┐
        q_2: ┤ P(-3) ├───────────────────────────┤ X ├┤ P(3) ├┤ X ├
             └───────┘                           └───┘└──────┘└───┘
```
Furthermore, the template shown above only produces one match on the circuit below.
```
             ┌───────┐
        q_0: ┤ P(-2) ├──■────────────■────────────────────────────
             ├───────┤┌─┴─┐┌──────┐┌─┴─┐┌──────┐
        q_1: ┤ P(-2) ├┤ X ├┤ P(2) ├┤ X ├┤ P(3) ├──■────────────■──
             └┬──────┤└───┘└──────┘└───┘└──────┘┌─┴─┐┌──────┐┌─┴─┐
        q_2: ─┤ P(3) ├──────────────────────────┤ X ├┤ P(3) ├┤ X ├
              └──────┘                          └───┘└──────┘└───┘
```

### CODE REVIEW

File: qiskit/circuit/instruction.py [hunk 1]
```diff
@@ -129,6 +129,52 @@ def __eq__(self, other):
...
+                except TypeError:
+                    pass
+
+                try:
+                    if numpy.isclose(float(self_param), float(other_param), atol=_CUTOFF_PRECISION):
+                        continue
+                except TypeError:
+                    pass
+            else:
+                continue
```
Comment (lcapelluto) [inline qiskit/circuit/instruction.py:172]: This is kind of hard to follow, is this the same?

```suggestion
            if isinstance(self_param, ParameterExpression) or \
                    isinstance(other_param, ParameterExpression):
                continue
            elif isinstance(self_param, numpy.array) and \
                    isinstance(other_param, numpy.array):
                if numpy.shape(self_param) == numpy.shape(other_param) \
                        and numpy.allclose(self_param, other_param, atol=_CUTOFF_PRECISION):
                    continue
            else:
                try:
                    if numpy.isclose(self_param, other_param, atol=_CUTOFF_PRECISION):
                        continue
                except TypeError:
                    pass
```

File: test/python/circuit/test_instructions.py [hunk 2]
```diff
@@ -84,6 +84,23 @@ def test_instructions_equal_with_parameters(self):
...
+            Instruction('u', 1, 0, [theta, phi, 0.4])))
+
+        # Verify that normal equality still holds.
+        self.assertTrue(Instruction('u', 1, 0, [0.4, 0.5]).soft_compare(
+            Instruction('u', 1, 0, [0.4, 0.5])))
+
+        # Test that when names differ we get False.
+        self.assertFalse(Instruction('u', 1, 0, [0.4, phi]).soft_compare(
+            Instruction('v', 1, 0, [theta, phi])))
+
```
Comment (lcapelluto) [inline test/python/circuit/test_instructions.py:103]: Please add a test that invokes the cutoff precision check in soft compare

**Review by lcapelluto — COMMENTED** [review]
This is the first half of my review, I'll look at the remaining few files tomorrow

File: qiskit/transpiler/passes/optimization/template_matching/template_substitution.py [hunk 3]
```diff
@@ -27,8 +31,9 @@ class SubstitutionConfig:
     in the circuit.
     """
 
-    def __init__(self, circuit_config, template_config, pred_block,
+    def __init__(self, template_dag_dep, circuit_config, template_config, pred_block,
```
[the remark this answers is not in the document]
Comment (eggerdj) [reply qiskit/transpiler/passes/optimization/template_matching/template_substitution.py:34]: No reason. I'll move it here: `qubit_config, template_dag_dep, clbit_config=None`.

File: qiskit/transpiler/passes/optimization/template_matching/template_substitution.py [hunk 4]
```diff
@@ -343,19 +355,19 @@ def run_dag_opt(self):
                 # Then add the inverse of the template.
                 for index in template_inverse:
                     all_qubits = self.circuit_dag_dep.qubits
-                    qarg_t = self.template_dag_dep.get_node(index).qindices
```
[the remark this answers is not in the document]
Comment (eggerdj) [reply qiskit/transpiler/passes/optimization/template_matching/template_substitution.py:346]: This was not a bug. This change is intentional and needed. Before, when the template was not parameterized, the template would never change (i.e. you can use `self`). Now, the template changes in the sense that the parameters can have different values. Consider the following template: `p(ϴ) p(-ϴ)`. The soft match will match with any two phase gates that follow, e.g. `p(2)p(2)`, `p(-1)p(1)` and `p(3)p(-4)` are examples of matches. However, only `p(-1)p(1)` is a valid match. To find out what is valid `_fake_bind` solves a system of equations for each tentative match, e.g. `ϴ=2, -ϴ=2`, `ϴ=-1, -ϴ=1`, and `ϴ=3, -ϴ=-4` for the previous examples. If a solution can be found the match is valid and values are assigned to a new template stored in `group` which will be used for the substitution. Hence the replacement `self`->`group`. This was not needed before.

File: qiskit/circuit/instruction.py [hunk 5]
```diff
@@ -129,6 +129,52 @@ def __eq__(self, other):
...
+
+        Returns:
+            bool: are self and other equal up to parameter expressions.
+        """
+        if self.name != other.name or \
+                other.num_qubits != other.num_qubits or \
+                other.num_clbits != other.num_clbits:
+            return False
+
+        for self_param, other_param in zip_longest(self.params, other.params):
```
[the remark this answers is not in the document]
Comment (eggerdj) [reply qiskit/circuit/instruction.py:150]: This is intended to check that the number of parameters match, right? And here we can understand that, e.g. `p(1)` and `p(ϴ)` both have one parameter?

File: qiskit/circuit/instruction.py (hunk 1 above)
Comment (eggerdj) [reply qiskit/circuit/instruction.py:172]: I think this is the same. I'll commit the suggestion and run the tests.

File: test/python/circuit/test_instructions.py (hunk 2 above)
Comment (eggerdj) [reply test/python/circuit/test_instructions.py:103]: Sure but I think we may want to be a bit careful here. On my machine this
```
aa = np.array([1., 0.40000+1e-6])
bb = np.array([1., 0.40000])
np.allclose(aa, bb, atol=1e-10)
```
 gives me `True`. But this
```
aa = np.array([1., 0.40000+1e-5])
bb = np.array([1., 0.40000])
np.allclose(aa, bb, atol=1e-10)
```
gives me `False`. This does not quite line up with `_CUTOFF_PRECISION = 1E-10`.

Comment (lcapelluto) [reply test/python/circuit/test_instructions.py:103]: Can we make it `+1e-20` for True and `+0.01` for false?

File: qiskit/circuit/instruction.py (hunk 1 above)
Comment (lcapelluto) [reply qiskit/circuit/instruction.py:173]: Right, it's a bit pseudo code-y. The correction is `isinstance(aa, np.ndarry)`

Comment (lcapelluto) [reply qiskit/circuit/instruction.py:173]: I can update my comment

File: qiskit/transpiler/passes/optimization/template_matching/template_substitution.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lcapelluto) [reply qiskit/transpiler/passes/optimization/template_matching/template_substitution.py:346]: I follow now. Thank you!:)

File: qiskit/transpiler/passes/optimization/template_matching/template_substitution.py [hunk 6]
```diff
@@ -375,3 +387,89 @@ def run_dag_opt(self):
 
         self.dag_dep_optimized = dag_dep_opt
         self.dag_optimized = dagdependency_to_dag(dag_dep_opt)
+
+    def _fake_bind(self, template_sublist, circuit_sublist):
+        """
+        Copy's the template and fake binds any parameters.
+        template_sublist and circuit_sublist match up to the
+        assignment of the parameters. For example the template
```
Comment (lcapelluto) [inline qiskit/transpiler/passes/optimization/template_matching/template_substitution.py:395]: Nice example

**Review by lcapelluto — CHANGES_REQUESTED** [review]
I think there's some room for making `soft_compare` more readable. This is very cool! Great docstrings especially, great examples

File: qiskit/circuit/instruction.py (hunk 1 above)
Comment (eggerdj) [reply qiskit/circuit/instruction.py]: Thanks that seems to work now. The soft_compare tests and template matching tests also pass.

File: qiskit/transpiler/passes/optimization/template_matching/template_substitution.py (hunk 6 above)
Comment (eggerdj) [reply qiskit/transpiler/passes/optimization/template_matching/template_substitution.py:395]: Thanks.

**Review by lcapelluto — DISMISSED** [review]

File: qiskit/circuit/instruction.py [hunk 7]
```diff
@@ -129,6 +129,45 @@ def __eq__(self, other):
...
+                other.num_qubits != other.num_qubits or \
+                other.num_clbits != other.num_clbits or \
+                len(self.params) != len(other.params):
+            return False
+
+        for self_param, other_param in zip_longest(self.params, other.params):
+            if isinstance(self_param, ParameterExpression) or \
+                    isinstance(other_param, ParameterExpression):
+                continue
+            elif isinstance(self_param, numpy.ndarray) and \
```
Comment (lcapelluto) [inline qiskit/circuit/instruction.py:155]: Lint is failing because this can be
```suggestion
            if isinstance(self_param, numpy.ndarray) and \
```

File: test/python/transpiler/test_template_matching.py [hunk 8]
```diff
@@ -202,6 +206,99 @@ def test_pass_template_wrong_type(self):
...
+            qc.cx(0, 1)
+            qc.p(beta, 1)
+            qc.cx(0, 1)
+            qc.append(CZp(2, [beta]), [0, 1])
+
+            return qc
+
+        def count_cx(qc):
+            """Counts the number of CX gates for testing."""
+            return sum([1 if inst[0].name == 'cx' else 0 for inst in qc.data])
```
Comment (ajavadia) [inline test/python/transpiler/test_template_matching.py:260]:
```suggestion
        def count_cx(qc):
            """Counts the number of CX gates for testing."""
            return qc.count_ops().get('cx', 0)
```

File: qiskit/transpiler/passes/optimization/template_matching/template_substitution.py [hunk 9]
```diff
@@ -375,3 +387,89 @@ def run_dag_opt(self):
...
+        q_1: ┤ P(-1.0*β) ├┤ X ├┤ P(β) ├┤ X ├┤1       ├
+             └───────────┘└───┘└──────┘└───┘└────────┘
+        should only match once in the circuit
+             ┌───────┐
+        q_0: ┤ P(-2) ├──■────────────■────────────────────────────
+             ├───────┤┌─┴─┐┌──────┐┌─┴─┐┌──────┐
+        q_1: ┤ P(-2) ├┤ X ├┤ P(2) ├┤ X ├┤ P(3) ├──■────────────■──
+             └┬──────┤└───┘└──────┘└───┘└──────┘┌─┴─┐┌──────┐┌─┴─┐
+        q_2: ─┤ P(3) ├──────────────────────────┤ X ├┤ P(3) ├┤ X ├
+              └──────┘                          └───┘└──────┘└───┘
```
Comment (ajavadia) [inline qiskit/transpiler/passes/optimization/template_matching/template_substitution.py:408]: I think you want to say "maximally match" here because even in the P(3) case there's still a match of CX-P(3)-CX with the middle part of the template.

File: qiskit/transpiler/passes/optimization/template_matching/template_substitution.py [hunk 10]
```diff
@@ -375,3 +387,89 @@ def run_dag_opt(self):
 
         self.dag_dep_optimized = dag_dep_opt
         self.dag_optimized = dagdependency_to_dag(dag_dep_opt)
+
+    def _fake_bind(self, template_sublist, circuit_sublist):
+        """
+        Copies the template and fake binds any parameters.
```
Comment (ajavadia) [inline qiskit/transpiler/passes/optimization/template_matching/template_substitution.py:393]:
```suggestion
        Copies the template and fake binds any parameters,
        i.e. attempts to solve for a valid parameter assignment.
```

File: qiskit/transpiler/passes/optimization/template_matching/template_substitution.py [hunk 11]
```diff
@@ -375,3 +387,89 @@ def run_dag_opt(self):
...
+        solution is found then the match is valid and the parameters
+        are assigned. If not, None is returned.
+
+        Args:
+            template_sublist (list): part of the matched template.
+            circuit_sublist (list): part of the matched circuit.
+
+        Returns:
+            DAGDependency: A deep copy of the template with
+                the parameters bound.
```
Comment (ajavadia) [inline qiskit/transpiler/passes/optimization/template_matching/template_substitution.py:425]:
```suggestion
        Returns:
            DAGDependency: A deep copy of the template with
                the parameters bound. If no binding satisfies the
                parameter constraints, returns None.
```

File: qiskit/transpiler/passes/optimization/template_matching/template_substitution.py [hunk 12]
```diff
@@ -375,3 +387,89 @@ def run_dag_opt(self):
 
         self.dag_dep_optimized = dag_dep_opt
         self.dag_optimized = dagdependency_to_dag(dag_dep_opt)
+
+    def _fake_bind(self, template_sublist, circuit_sublist):
```
Comment (ajavadia) [inline qiskit/transpiler/passes/optimization/template_matching/template_substitution.py:391]: "fake_bind" might imply that it doesn't actually bind, but in fact it does bind if some valid params are found. So maybe a different name for this function? something like `_attempt_bind`?

Comment (lcapelluto) [reply qiskit/transpiler/passes/optimization/template_matching/template_substitution.py:391]: I don't think it modifies `self`, though, right? Does that distinction matter here?

Comment (eggerdj) [reply qiskit/transpiler/passes/optimization/template_matching/template_substitution.py:391]: No, it does not modify `self`. That being said, I like `_attempt_bind` as a bind is not guaranteed.

File: qiskit/transpiler/passes/optimization/template_matching/template_substitution.py (hunk 9 above)
Comment (eggerdj) [reply qiskit/transpiler/passes/optimization/template_matching/template_substitution.py]: Yes, I amended the docstring.

**Review by lcapelluto — APPROVED** [review]

**Review by ajavadia — APPROVED** [review]
